File: javatest/src/main/java/examples/algorithms/BinarySearch.java
package examples.algorithms;

/**
 * Что такое Binary Search?
 * <p>
 * Binary Search (Двоичный поиск) — это эффективный алгоритм для поиска элемента в отсортированном массиве или списке. Алгоритм работает следующим образом:
 * <p>
 * - Мы начинаем с середины массива.
 * <p>
 * - Если элемент в середине массива — это искомый элемент, то поиск завершён.
 * <p>
 * - Если искомый элемент меньше среднего элемента, то мы повторяем поиск в левой половине массива.
 * <p>
 * - Если искомый элемент больше среднего элемента, то мы ищем в правой половине массива.
 * <p>
 * - Этот процесс повторяется, пока не найдём элемент или не определим, что его нет в массиве.
 * <p>
 * Важные особенности:
 * <p>
 * Массив должен быть отсортирован.
 * Алгоритм работает за O(log N), что делает его очень быстрым для больших массивов.
 * <p>
 * <br/>
 * Пример задачи:
 * Дан массив чисел и искомое число target, найти это число в массиве и вернуть его индекс
 */
public class BinarySearch {
    public static int search(int[] nums, int target) {
        int left = 0, right = nums.length - 1;

        while (left <= right) {
            int mid = left + ((right - left) / 2);

            if (nums[mid] == target) {
                return mid; // Element found
            } else if (nums[mid] < target) {
                left = mid + 1; // Search right half
            } else {
                right = mid - 1; // Search left half
            }
        }

        return -1; // Element not found
    }

    public static void main(String[] args) {
        var arr = new int[]{1, 2, 3, 4, 5, 6, 7, 8, 9, 10};
        System.out.println(search(arr, 2));
        System.out.println(search(arr, 5));
        System.out.println(search(arr, 8));
        System.out.println(search(arr, 1));
        System.out.println(search(arr, 10));
    }
}

File: _python_/algorithms/binary_search.py
def binary_search(arr, n):
    left = 0
    right = len(arr) - 1
    while left <= right:
        mid = (left + right) // 2
        if n < arr[mid]:
            right = mid - 1
        elif arr[mid] < n:
            left = mid + 1
        else:
            return mid
    return -1


def binary_search_recursive(arr, n, left=0, right=None):
    """Рекурсивный бинарный поиск. Возвращает индекс элемента n в массиве arr или -1, если элемент не найден."""
    if right is None:
        right = len(arr) - 1  # Устанавливаем предел при первом вызове

    if left > right:
        return -1  # Элемент не найден

    mid = (left + right) // 2  # Находим середину

    if arr[mid] == n:
        return mid  # Найден элемент, возвращаем индекс
    elif arr[mid] < n:
        return binary_search_recursive(arr, n, mid + 1, right)  # Поиск в правой половине
    else:
        return binary_search_recursive(arr, n, left, mid - 1)  # Поиск в левой половине


arr = [1, 3, 5, 7, 9, 11]

print(binary_search(arr, 5))  # 2
print(binary_search(arr, 8))  # -1
print(binary_search(arr, 1))  # 0
print(binary_search(arr, 11))  # 5
print(binary_search(arr, 0))  # -1

print()

print(binary_search_recursive(arr, 5))  # 2
print(binary_search_recursive(arr, 8))  # -1
print(binary_search_recursive(arr, 1))  # 0
print(binary_search_recursive(arr, 11))  # 5
print(binary_search_recursive(arr, 0))  # -1
